man/man1/diversity.1
\"
\"
\"
\"
\"
\"
\"
.Dd 07 July 2007
.Os
.Dt DIVERSITY 1
.Sh NAME
.Nm diversity
.Nd "compute diversity statistics for an ensemble"
.Sh SYNOPSIS
.Nm
--format=NAME
format-specific-arguments
.Sh DESCRIPTION
.Nm
operates on exodus\- or avatar\-format data and an ensemble file to compute
diversity statistics.
.Pp
The compute statistics are:
.Bl -tag -width "--concatenate" -compact
.It 1) Average Dietterich Kappa
.It 2) Pair-wise Q Statistic
.It 3) Inter-rater agreement function (Kappa)
.It 4) Percentage correct diversity measure (PDCM)
.El
.Pp
In addition, a plot file can be generated listing the Dietterich Kappa and
average error rate for each pair of classifiers.
.Sh ARGUMENTS AND OPTIONS
.Ss REQUIRED ARGUMENTS
.Pp
.Bl -tag -width "--concatenate" -compact
.It Fl o Ar format-type
.It Fl -format Ar format-type
.Ar format-type
is either
.Ar exodus
or
.Ar avatar .
See the
.Sx FORMATS
section for details. The default is
.Ar exodus
.El
.Pp
NOTE: See the
.Sx FILENAMES
section for file naming conventions.
.Pp
.so man1/exodus_args
.El
.Pp
.so man1/avatar_args
.Pp
.so man1/avatar_opts
.Pp
.Ss ALTERNATE FILENAMES
.Pp
.Bl -tag -width "--concatenate" -compact
.It Fl -test-file= Ar file
For
.Ar avatar
data, override the default filename for testing data.
.It Fl -namesfile= Ar file
For
.Ar avatar
data, override the default filename for metadata (the
.Ar names
file.
.It Fl -trees-file= Ar file
Override the default filename for the ensemble of trees.
.El
.Pp
.so man1/formats
.Pp
.Sh FILENAMES
.Pp
For
.Ar exodus
format data, if the value of the
.Fl -datafile
argument is
.Ar foo.ex2
then the testing data is taken from
.Ar foo.ex2
and the ensemble is read from
.Ar foo.trees .
.Pp
For
.Ar avatar
format data, if the value of the
.Fl -filestem
argument is
.Ar bar
then
.Ar bar.names
is used for the class and attribute definition file,
.Ar bar.test
is used for the testing data.
The ensemble is read from
.Ar bar.trees .
.Pp
.Sh EXAMPLES
.Pp
.Dl diversity -o avatar -f autos --truth-column=3 --exclude=1-2
.Pp
Will compute the diversity statistics from the ensemble "autos.trees" using the data in
"autos.test". The truth column is the third column and the first two columns are ignored.
.Pp
.Dl diversity -o avatar -f autos --truth-column=26 --output-kappa-plot-data
.Pp
Will do the same as the example above plus it will print a plot file containing four columns:
.Bl -tag -width "--concatenate" -compact
.It 1) classifier 1 index
.It 2) classifier 2 index
.It 3) Kappa value for classifiers 1 and 2
.It 4) average accuracy for classifiers 1 and 2
.El
.Pp
.Dl avatardt -o exodus -d can.ex2 -C can.classes --test-times=20-24
.Pp
Will compute the diversity statistics from the ensemble file "can.trees" using the data
in "can.ex2" for timesteps 20-24.
.Sh FILES
.Pp
.Pp
.Sh SEE ALSO
.Pp
.Xr avatardt 1,
.Xr crossvalfc 1
.Pp
.Sh HISTORY
.Pp


ChangeLog
ADDED FEATURES:

The installation process is now
./configure && make && make install

Added the --verbose-oob option to avatardt. This generates a new file
(default is basename.oob-data) containing the OOB error as each new
tree is added to the ensemble. Requires that either bagging or ivoting
be turned on.

Added the --version option which does what you may expect.

You can now specify alternate filenames for input/output instead of
being forced to use the same basename for all. See the --train-data
option and others like it in the usage message or manpage.

The overall voted accuracy is printed when running crossvalfc even
when the confusion matrix is not requested.

BUG FIXES:

Fixed an --include bug that manifests when truth is column 1 and you
want to include above column 2.

Fixed a crash when reading some trees due to memory problems.

Fixed a problem with empty files specified as training/testing data.
The code now alerts you to this fact and quits instead of crashing.

Fixed a bug that didn't like some comment lines in the training/
testing data.

OTHER:

The version is now integrated with the latest FCLib release (1.6.1)
which has had all SAF references removed. It is also integrated with
the latest ExodusII open source release (4.61).


tools/diversity.c
/********************************************************************************** 
Avatar Tools 
Copyright (c) 2019, National Technology and Engineering Solutions of Sandia, LLC
All rights reserved. 

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright
  notice, this list of conditions and the following disclaimer  in the
  documentation and/or other materials provided with the distribution.


3. Neither the name of the copyright holder nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

For questions, comments or contributions contact 
Philip Kegelmeyer, wpk@sandia.gov 
*******************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "../src/crossval.h"
#include "../src/options.h"
#include "../src/version_info.h"
#include "../src/rw_data.h"
#include "../src/tree.h"
#include "diversity_measures.h"

int main(int argc, char **argv) {
    DT_Ensemble Ensemble;
    Args_Opts Args;
    AV_SortedBlobArray Sorted_Examples;
    FC_Dataset ds;
    CV_Dataset Dataset;
    CV_Subset Subset;
    int **result_matrix;
    
    Args = process_opts(argc, argv);
    Args.do_testing = TRUE;
    Args.caller = DIVERSITY_CALLER;
    if (! sanity_check(&Args))
        exit(-1);
    set_output_filenames(&Args, FALSE, FALSE);
    
    if (Args.format == EXODUS_FORMAT && ! Args.do_training) {
        init_fc(Args);
        open_exo_datafile(&ds, Args.datafile);
    }
    av_exitIfError(av_initSortedBlobArray(&Sorted_Examples));
    read_testing_data(&ds, Subset.meta, &Dataset, &Subset, &Sorted_Examples, &Args);
    late_process_opts(Dataset.meta.num_attributes, Dataset.meta.num_examples, &Args);
    
    read_ensemble(&Ensemble, -1, 0, &Args);
    
    // Initialize matrix holding per-tree results
    init_matrix(Ensemble, Subset, &result_matrix);
    printf("Average Kappa     = %f\n", compute_dietterich_kappa(Ensemble.num_trees, Subset.meta.num_classes,
                                                                Subset.meta.num_examples, result_matrix, Args));
    printf("Q Statistic       = %f\n", compute_Q_statistic(Ensemble.num_trees, Subset.meta.num_examples,
                                                           result_matrix));
    printf("Inter-rater Kappa = %f\n", compute_interrater_kappa(Ensemble.num_trees, Subset.meta.num_examples,
                                                                result_matrix));
    printf("PCDM              = %f\n", compute_pcdm(Ensemble.num_trees, Subset.meta.num_examples, result_matrix));
    return 0;
}

void display_usage( void ) {
    printf("\ndiversity ");
    printf(get_version_string());
    printf("\n");
    printf("\nUsage: diversity options\n");
    printf("\nbasic arguments:\n");
    printf("    -o, --format=FMTNAME     : Data format. FMTNAME is either 'exodus' or 'avatar'\n");
    printf("                               Default = 'exodus'\n");
    printf("    --output-kappa-plot-data : Write a plot file containing the following four columns:\n");
    printf("                               1) classifier 1 number\n");
    printf("                               2) classifier 2 number\n");
    printf("                               3) Kappa value for the two classifiers\n");
    printf("                               4) Average error for the two classifiers\n");
    printf("\n");
    printf("required exodus-specific arguments:\n");
    printf("    -d, --datafile=FILE     : Use FILE as the exodus datafile\n");
    printf("    --test-times=R          : Use data from the range of times R to test (e.g. 5,7-10)\n");
    printf("    -V, --class-var=VARNAME : Use the variable named VARNAME as the class\n");
    printf("                              definition\n");
    printf("    -C, --class-file=FILE   : FILE the gives number of classes and thresholds:\n");
    printf("                              E.g.\n");
    printf("                                  class_var_name Osaliency\n");
    printf("                                  number_of_classes 5\n");
    printf("                                  thresholds 0.2,0.4,0.6,0.8\n");
    printf("                                Will put all values <=0.2 in class 0,\n");
    printf("                                <=0.4 in class 1, etc\n");
    printf("\n");
    printf("required avatar-specific arguments:\n");
    printf("    -f, --filestem=STRING : Use STRING as the filestem\n");
    printf("\n");
    printf("avatar-specific options:\n");
    printf("    --include=R      : Include the features listed in R (e.g. 1-4,6)\n");
    printf("    --exclude=R      : Exclude the features listed in R (e.g. 1-4,6)\n");
    printf("                       --include and --exclude may be specified multiple times.\n");
    printf("                       They are applied left to right.\n");
    printf("    --truth-column=S : Location of the truth column. S = first or last\n");
    printf("                       Default = last\n");
    printf("\n");
    printf("alternate filenames:\n");
    printf("    --names-file=FILE       : For avatar format data, use FILE for the names file\n");
    printf("    --test-file=FILE        : For avatar format data, use FILE for testing data\n");
    printf("    --trees-file=FILE       : For avatar format data, use FILE for ensemble file\n");
    printf("\n");
    exit(-1);
}

